Fixed issue - #19346 Import data 2.2.6 Value for 'product_type' attribute contains incorrect value

GovindaSharma opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
Fixed issue - #19346
Import data 2.2.6 Value for 'product_type' attribute contains incorrect value

Since 'product_attribute' is reserved attribute,so we can restrict this type of attribute to be created by user.

### Fixed Issues (if relevant)

Fixed issue - #19346
Import data 2.2.6 Value for 'product_type' attribute contains incorrect value

### Manual testing scenarios (*)

   1. create an attribute with code product_type http://i.prntscr.com/AwUJuhizQKatBwNwRa1S7Q.png with type: dropdown scope: global sortable, comparable: no
  2.  add it to a family and create a product.
   3. go to import profiles download sample file then reimport sample file


### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Fixed issue related to 'product_attribute' attribute.
Added condition for validating product attribute.

orlangur commented on app/code/Magento/Catalog/Controller/Adminhtml/Product/Attribute/Validate.php:
```diff
@@ -115,6 +115,14 @@ public function execute()
+    		$message = 	__('Cannot create attribute (%1) as it is already used in catalog',$attributeCode);
+    		$this->setMessageToResponse($response, [$message]);
+    		$response->setError(true);
+    		$response->setProductAttribute($attribute->toArray());
```
Please add a proper check into

```
if ($attribute->getId() && !$attributeId) {
            $message = strlen($this->getRequest()->getParam('attribute_code'))
                ? __('An attribute with this code already exists.')
                : __('An attribute with the same code (%1) already exists.', $attributeCode);
```

condition, do not introduce a constant - just compare with `'product_type'` string literal.

GovindaSharma replied:
so instead of introducing a constant...i should compare the string..if string is similar then throw msg right?

orlangur replied:
@GovindaSharma exactly, no need to overcomplicate things. When we need more than one such forbidden not-really-an-attribute, we may think about array.

orlangur replied:
@GovindaSharma,

```
if(strcmp($attributeCode,'product_type')==false){
    		$message = 	__('Cannot create attribute (%1) as it is already used in catalog',$attributeCode);
```

still not combined with first condition as requested. `$attributeCode !== 'product_type'` should do the trick.

GovindaSharma replied:
hi @orlangur 

"Please fill all description fields according to template and apply code changes, then squash it into single commit and force push."

this was your feedback..i did not understand it.can you please tell me give any idea what changes i need to apply in code?

orlangur replied:
@GovindaSharma check this: https://github.com/magento/magento2/pull/19408/files#r236223352

You need to add new check into existing `if` instead of introducing a new one.

orlangur replied:
@GovindaSharma still not fixed

orlangur reviewed (changes requested):
Please fill all description fields according to template and apply code changes, then squash it into single commit and force push.

GovindaSharma commented:
changes done as per request

GovindaSharma commented:
@orlangur can you please tell me what is the exact issue in first check which is not passed?i have tested the code in php 7.2 version.Please let me the exact issue so that i can fix the same

GovindaSharma commented:
since two checks has been passed but one is not..so what issue occurred in that checkes for which it has not been passed?

GovindaSharma commented:
@orlangur please look into the changes and let me know what other things i need to do?

GovindaSharma commented:
@magento-engcom-team and @orlangur ,can you please review the changes as i have done all necessary changes suggested by @orlangur.Let me know your feedback on this.

orlangur commented on app/code/Magento/Catalog/Model/ResourceModel/Eav/Attribute.php:
```diff
@@ -37,7 +37,6 @@ class Attribute extends \Magento\Eav\Model\Entity\Attribute implements
 
     const KEY_IS_GLOBAL = 'is_global';
-
     /**
```
Revert this change)

GovindaSharma replied:
are you talking about the extra space in file? that i need to remove?

GovindaSharma replied:
@orlangur .i made the chnages.Please review this.
Thank You.

orlangur replied:
Try `git checkout 2.3-develop -- app/code/Magento/Catalog/Model/ResourceModel/Eav/Attribute.php` to get this file in correct state.

GovindaSharma replied:
@orlangur file is in correct state now.I made the changes.

GovindaSharma commented:
@magento-engcom-team and @orlangur .i have done changes requested by @orlangur .Please review this and let me know.

orlangur commented:
@GovindaSharma please check https://github.com/magento/magento2/pull/19408/files#r237182111. All needed changes are not done yet.

GovindaSharma commented:
@orlangur , now i have added check into existing if,previously i added extra if just show different message because product_type in real does not exist in magento,so that was the reason to add extra if.But now i added the checks into existing if condition.Please review and let me know.

Thank You

orlangur commented:
@GovindaSharma thanks! Looks good now, please squash changes into a single commit (incuding my commits) and we are good to go.

GovindaSharma commented:
@orlangur .I have now squash all commits into one and done force push.I think all the changes has been done now.Please check and approve the changes.
Thank You

magento-engcom-team commented:
Hi @orlangur, thank you for the review.
[ENGCOM-3597](https://jira.corp.magento.com/browse/ENGCOM-3597) has been created to process this Pull Request

magento-engcom-team commented:
@GovindaSharma thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

magento-engcom-team commented:
Hi @GovindaSharma. Thank you for your contribution. 
We will aim to release these changes as part of 2.3.1. 
Please check the release notes for final confirmation.

GovindaSharma commented:
@magento-engcom-team ,thanks for your confirmation.
